### ISSUE TITLE
Download passwords

### ISSUE THREAD

**Issue by tire-fire** [HaveIBeenPwned/PwnedPasswordsDownloader#50]
This just downloads password hashes, not the passwords themselves. I'm trying to find the passwords that were used to generate these hashes. Are the passwords available for download somewhere? Calling this a PasswordDownloader is misleading.

**Comment by troyhunt** [comment]
You can't download the plain text passwords, here's why: https://www.troyhunt.com/understanding-have-i-been-pwneds-use-of-sha-1-and-k-anonymity/

**Comment by quantumpacket** [comment]
@tire-fire download `Collections #1-5` as a starting point if you want plain-text passwords and user/password dumps for data analysis and machine learning training. You can use https://github.com/p3pperp0tts/leaks_parser to deduplicate and consolidate them. Note, you will need at least a 500GB drive.

**Comment by troyhunt** [comment]
> I see, I think you can't share the password lists that you have because hackers are bad at parsing and you can't trust them not to have PII in their data. This is done at the risk of having false positives which may happen in the form of hash collisions with non-password data.

It's not just parsing, some people deliberately put PII in their passwords. Hash collisions are practically impossible; of the more than 500M passwords loaded into HIBP before the ingestion pipeline was launched, no 2 hashes collided. With 16^40 different possible SHA-1 hashes and less than 1B passwords, it is for all intents and purposes near impossible to have a collision. If a collision occurs because a password is tested that results in the same hash already in the system, the impact of the false positive is merely that someone is told a password has been seen before when, in fact, it hasn't.

In summary, there is an infinitesimally small chance of a collision and the impact if that happens is of absolutely no consequence.

**Comment by tire-fire** [comment]
Thanks, but I still suggest a rename to PasswordHashDownloader, because there are no passwords downloaded by this software.

**Comment by troyhunt** [comment]
There are passwords downloaded. They’re stored as hashes. There’s so much extensive information out there about precisely how this service works, this is honestly the first time we’ve had this suggestion.

**Comment by tire-fire** [comment]
Maybe I'm alone in this thinking, but I would say that storing something in an unrecoverable format isn't a storage mechanism for the thing. So I would say there are no passwords downloaded, probably as a feature that you intended.

I apologize as I've never used this service before, but I was looking for a list of actual passwords and a PwnedPasswordDownloader sounds like it would fit what I was looking for. And I see now that the README says that it is just hashes that are downloaded.

**Comment by troyhunt** [comment]
> And I see now that the README says that it is just hashes that are downloaded.

😊
